### ISSUE TITLE
H2TableDefinition with H2 2.0+ produces wrong code generation output when the run mode is Oracle

### ISSUE THREAD

**Issue by mglazer** [jOOQ/jOOQ#13302]
### Expected behavior

When querying the schema for an H2 table (in order to use a Generator against an H2 database), I expect that all types are appropriately determined.

### Actual behavior

It seems that all types are determined to be array types. I believe this is because of an incorrect `nvl` statement in the `H2TableDefinition`:

https://github.com/jOOQ/jOOQ/blob/c3a739680b093870595699e46c773fee542126ef/jOOQ-meta/src/main/java/org/jooq/meta/h2/H2TableDefinition.java#L114-L117

If my understanding of `nvl` is correct, then it picks the first non-null value. However, because of the `concat` with `" ARRAY"` the first value is always non-null, and therefore all types get labeled as arrays.

I've verified this by just setting up break points and seeing that no matter what I do, the definedType set in the DefaultColumnDefinition is set to:

```
DataType [ t=ARRAY]
```

And the generated code always looks like the following:

```
    /**
     * The column <code>PUBLIC.INVENTORY.ID</code>.
     */
    public final TableField<InventoryRecord, Object[]> ID = createField(DSL.name("ID"), org.jooq.impl.DefaultDataType.getDefaultDataType("\"\"").getArrayDataType(), this, "");
```

### Steps to reproduce the problem

Run `GenerationTool` against any H2 2.0+ database.

### Versions

- jOOQ: 3.16.3 (planning to upgrade to 3.16.5 as soon as I can)
- Java: 11
- Database (include vendor): H2 embedded 2.1.210
- OS: OSX
- JDBC Driver (include name if inofficial driver): H2

I'm happy to commit a fix here, I believe it's just reversing the order of arguments in the nvl.

**Comment by mglazer** [comment]
Proposal: https://github.com/jOOQ/jOOQ/pull/13303

**Comment by lukaseder** [comment]
Thanks for your report. Our integration tests suggest that everything is fine. The only reason why this particular `NVL()` expression could behave differently was if H2 was run in some compatibility run mode, where concatenating `NULL || ' ARRAY'` produces `' ARRAY'` instead of standard SQL `NULL`. I'm guessing, the run mode could be Oracle?

While I don't think using these run modes is a good idea when using jOOQ's code generator, we could implement an explicit `NULL` check rather than using the current implicit one. I'll create a patch that you could verify on your end.

Alternatively, if my hypothesis about run modes is wrong, I'd appreciate a more complete reproducer. As I said, our own integration tests don't run into any such problem.

**Comment by lukaseder** [comment]
So, this would be the patch:

```patch
diff --git a/jOOQ-meta/src/main/java/org/jooq/meta/h2/H2TableDefinition.java b/jOOQ-meta/src/main/java/org/jooq/meta/h2/H2TableDefinition.java
index ba4f652..da7f370 100644
--- a/jOOQ-meta/src/main/java/org/jooq/meta/h2/H2TableDefinition.java
+++ b/jOOQ-meta/src/main/java/org/jooq/meta/h2/H2TableDefinition.java
@@ -110,12 +110,12 @@
                 COLUMNS.COLUMN_NAME,
                 COLUMNS.ORDINAL_POSITION,
 
+                // [#13302] Use explicit NULL check to prevent side-effects from H2's compatibility run modes
+                when(ELEMENT_TYPES.DATA_TYPE.isNotNull(), concat(ELEMENT_TYPES.DATA_TYPE, inline(" ARRAY")))
                 // [#2230] [#11733] Translate INTERVAL_TYPE to supported types
-                nvl(concat(ELEMENT_TYPES.DATA_TYPE, inline(" ARRAY")),
-                    when(COLUMNS.INTERVAL_TYPE.like(any(inline("%YEAR%"), inline("%MONTH%"))), inline("INTERVAL YEAR TO MONTH"))
-                      .when(COLUMNS.INTERVAL_TYPE.like(any(inline("%DAY%"), inline("%HOUR%"), inline("%MINUTE%"), inline("%SECOND%"))), inline("INTERVAL DAY TO SECOND"))
-                      .else_(Tables.COLUMNS.DATA_TYPE)
-                ).as(COLUMNS.TYPE_NAME),
+                .when(COLUMNS.INTERVAL_TYPE.like(any(inline("%YEAR%"), inline("%MONTH%"))), inline("INTERVAL YEAR TO MONTH"))
+                .when(COLUMNS.INTERVAL_TYPE.like(any(inline("%DAY%"), inline("%HOUR%"), inline("%MINUTE%"), inline("%SECOND%"))), inline("INTERVAL DAY TO SECOND"))
+                .else_(Tables.COLUMNS.DATA_TYPE).as(COLUMNS.TYPE_NAME),
                 nvl(ELEMENT_TYPES.CHARACTER_MAXIMUM_LENGTH, COLUMNS.CHARACTER_MAXIMUM_LENGTH).as(COLUMNS.CHARACTER_MAXIMUM_LENGTH),
                 coalesce(
                     ELEMENT_TYPES.DATETIME_PRECISION,
```

Does this fix the problem on your end?

**Comment by katzyn** [comment]
> I'm guessing, the run mode could be Oracle?

You're right, H2 returns `' ARRAY'` from `NULL || ' ARRAY'` in Oracle compatibility mode only (because concatenation works in this way in Oracle).

**Comment by lukaseder** [comment]
In any case, the patched version of the query does not affect any generation output in our integration tests, so I'd say it's equivalent in normal run modes, while working around this caveat in the event of an Oracle run mode.

We don't support the latter, nor do we test it, but since this is a highly probable hypothesis (thanks, @katzyn), I'll just go ahead and change the query accordingly.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0 and 3.16.6 (https://github.com/jOOQ/jOOQ/issues/13305). A backport to 3.15.x isn't planned, as the query worked quite differently in 3.15 (not supporting H2 2.x yet).

**Comment by mglazer** [comment]
You’re way too fast. Your hypothesis was correct. I was running in ORACLE compatibility mode for H2 which is why I was hitting this. I didn’t realize that would have affected the results as much as they did. 

Thank you for the fix and teaching me something new.
